kubernetes/src/main/java/com/linecorp/armeria/client/kubernetes/endpoints/KubernetesEndpointGroup.java

```diff
@@ -266,50 +257,105 @@ public static KubernetesEndpointGroupBuilder builder(Config kubeConfig) {
         this.nodeAddressFilter = nodeAddressFilter;
         this.autoClose = autoClose;
         this.maxWatchAgeMillis = maxWatchAgeMillis == Long.MAX_VALUE ? 0 : maxWatchAgeMillis;
-        watchJob(this::watchNode);
-        watchJob(this::watchService);
+        executeJob(() -> start(true));
     }
 
-    private void watchService() {
-        watchLock.lock();
-        try {
-            final Watch oldServiceWatch = serviceWatch;
-            if (oldServiceWatch != null) {
-                oldServiceWatch.close();
-            }
+    private void start(boolean initial) {
+        if (wipUpdater.getAndIncrement(this) > 0) {
+            // Another thread is already starting.
+            return;
+        }
+
+        do {
             if (closed) {
                 return;
             }
-            final Watch newServiceWatch;
-            logger.info("[{}/{}] Start the service watcher...", namespace, serviceName);
-            try {
-                newServiceWatch = doWatchService();
-            } catch (Exception e) {
-                logger.warn("[{}/{}] Failed to start the service watcher.", namespace, serviceName, e);
+            if (!doStart(initial)) {
+                wipUpdater.set(this, 0);
                 return;
             }
-            // Recheck the closed flag because the doWatchService() method may take a while.
-            if (closed) {
-                newServiceWatch.close();
-                return;
+            initial = false;
+            // Repeat until all `start()` requests are handled.
```

Comment:
Can't it just to this only once when `start()` method is called concurrently? I'm not sure why it needs to handle all of them.

Reply:
It would be a rare case. `start()` can be called recursively if an exception occurs during execution `start()`.

Reply:
In that case, it still doesn't have to repeatedly call the `start()` method. Isn't that right?
Can't we just do something like:
```
private void start(boolean initial) {
    if (closed) {
        return;
    }

    if (wipUpdater.getAndIncrement(this) > 0) {
        // Another thread is already starting.
        wipUpdater.decrementAndGet(this);
        return;
    }

    doStart(initial);
    wipUpdater.decrementAndGet(this);
}
```

Reply:
> it still doesn't have to repeatedly call the start() method. Isn't that right?

That wouldn’t work. If start fails, a retry logic is triggerd. But the above code may disable that retry process.

```
STARTING 
-> FAIL 
-> RETRY in 0 ms for the first failure
-> RETRY invokes start() but being ignored by wipUpdater.decrementAndGet(this)
-> TERMINATED start() is terminated without retry
```

Reply:
Now I understand. Thanks for the explanation. 😉